File: 3-generate-config-files/generate_apsim_configs.py
#!/usr/bin/env python3

import os
import itertools
import csv

def generate_config_files(soil_names_file, weather_dir, base_config_file, output_dir):
    # Read soil names from the CSV file
    with open(soil_names_file, 'r', newline='') as f:
        csv_reader = csv.reader(f)
        soil_names = [row[0] for row in csv_reader if row]  # Assuming soil names are in the first column

    # Get weather files
    weather_files = [f.split('.')[0] for f in os.listdir(weather_dir) if f.endswith('.met')]

    # Read base config
    with open(base_config_file, 'r') as f:
        base_config = f.read()

    # Generate configurations
    for soil_name, weather_file in itertools.product(soil_names, weather_files):
        # Replace placeholders in the config
        config = base_config.replace('WeatherFileName', weather_file)
        config = config.replace('SoilName', soil_name)
        
        # Generate output filename
        output_filename = f"{weather_file}_{soil_name}.txt"
        output_path = os.path.join(output_dir, output_filename)

        # Write the config file
        with open(output_path, 'w') as f:
            f.write(config)

        print(f"Generated config file: {output_filename}")

# Usage
generate_config_files(
    soil_names_file='SubsetSoilName.csv',
    weather_dir='Weather',
    base_config_file='ExampleConfig.txt',
    output_dir='ConfigFiles'
)

File: 3-generate-config-files/generate_apsim_configs.R
#!/usr/bin/env Rscript
###Create config files


library(tidyr)
SoilNames <- as.vector(unlist(read.csv("SubsetSoilName.csv"))) ##List of soil names that must exist in the Soil Library .apsimx file (most likely supplied by experienced APSIM user)
WeatherFiles <- list.files('./', pattern = "\\.met$") %>% gsub('.met', '', .) ##Listing all available weather files.
ConfigTemplate <- "ExampleConfig.txt" ##Base config file

#Temporary Subset of soilNames
SoilNames <- SoilNames[1:10]


Des <- expand.grid(SoilNames, WeatherFiles) #make factorial design from factors given

DBNames <- vector() ##Object initiation
apsimxNames <- vector()

CreateConfigFiles <- function(ConfigTemplate, Des){
  for(i in 1:nrow(Des)){
    #Read Base config file and change variables to correct names
    Config <- readLines(ConfigTemplate)
    Config <- gsub(pattern = 'SoilName', replacement = Des$Var1[i], Config)
    Config <- gsub(pattern = 'WeatherFileName', replacement = Des$Var2[i], Config)

    #Write config information to a new txt file
    ConfigFilePath <- paste0("./", Des$Var1[i], "_", Des$Var2[i], "ConfigFile.txt")
    fileConn <- file(ConfigFilePath)
    writeLines(Config, fileConn)
    close(fileConn)

    #save db names in a vector so that they can be saved and loaded into python
    apsimxNames[i] <- paste0(Des$Var2[i], "_", Des$Var1[i], ".apsimx")
    DBNames[i] <- paste0(getwd(), "/", Des$Var2[i], "_", gsub(" ", "",Des$Var1[i]), '.db')
  }
  out <- data.frame(apsimxNames, DBNames)
  write.csv(out, file = "FileAndDBNames.csv")
}

CreateConfigFiles(ConfigTemplate, Des)
